File: TowerDefence.spritebuilder/Utils/StackObject.swift
//
//  StackObject.swift
//  TowerDefence
//
//  Created by koveres on 02.12.14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

import Foundation


public class StackObjectAlreadyEngaged:  NSException{}


public  class  StackObject: ObjectWithID{

    
    init(data :AnyObject){ userData = data;}
    
    var ID_Object:CInt = 0;
    var userData:AnyObject;
    
    var isEngaged = false;
    
    
    public  func getUserData() -> (AnyObject) {
        return self.userData;
    }
    
    public  func setUserData(data :AnyObject) {
         userData = data;
    }

    public func getIdObject() -> (CInt)
    {
        return ID_Object;
    }
    

    
    private struct lastId { static var id: CInt = 0 }
    
    private func nextID() -> (CInt){
        
        lastId.id++;
        if(lastId.id == CInt.max){
            lastId.id = 0;}
        return lastId.id;
    }
    
    
    
   public func UseStackObject()
   {
    if(isEngaged){
        StackObjectAlreadyEngaged().raise();
    }
    isEngaged = true;
        ID_Object = nextID();
    }
    
    
    
    public func isFree() -> (Bool)
    {
        return !isEngaged;
    }
    
    public func clearObject(){}
    
    public func FreeStackObject()
    {
        clearObject();
        isEngaged = false;
    }
    
}

File: test.spritebuilder/Source/TDEngine/Utils/StackObject.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Utils;

import internalObject.ObjectWithID;

/**
 *
 * @author Константин
 */

public abstract class  StackObject implements ObjectWithID{
    private int ID_Object;
    static int lastId = 0;
    
    Object userData;
    @Override
    public Object getUserData() {
        return this.userData;
    }

    @Override
    public void setUserData(Object data) {
          this.userData = data;
    }
    @Override
    public int getIdObject()
    {
        return this.ID_Object;
    }
    private boolean isEngaged = false;

    private int nextID() {
        lastId++;
        if(lastId == Integer.MAX_VALUE) 
            lastId = 0;
        return lastId;
    }

    public class StackObjectAlreadyEngaged extends  Exception{};
   
    
    public void UseStackObject() throws StackObjectAlreadyEngaged {
        if(isEngaged)
            throw new StackObjectAlreadyEngaged();
        isEngaged = true;
      
        this.ID_Object = nextID();
    }

    public boolean isFree()
    {
        return !isEngaged;
    }

    public abstract void  clearObject();

    public void FreeStackObject()
    {
        clearObject();
        isEngaged = false;
    }

}
